src/app/common/components/RouteGuard.tsx

```diff
@@ -21,23 +29,33 @@ const RouteGuard: React.FunctionComponent<IRouteGuard> = ({
   const [showRouteGuard, setShowRouteGuard] = React.useState(when);
   const [currentPath, setCurrentPath] = React.useState('');
 
+  const unblockFnRef = React.useRef<UnregisterCallback | null>(null);
+  const unblock = () => {
+    if (unblockFnRef.current) {
+      unblockFnRef.current();
+      unblockFnRef.current = null;
+    }
+  };
+
   React.useEffect(() => {
     if (when) {
-      history.block((prompt) => {
+      // When browser history routing happens (clicking links, back/forward) we can handle it with our custom modal.
+      unblockFnRef.current = history.block((prompt) => {
```

Comment:
Can you help me understand why storing this return value in a ref works, but setting a state property with the same callback doesn't? I'm sure it's just a gap in my understanding of the lifecycle differences between useRef and useState.. I was essentially trying..

```
const [unblockFn, setUnblockFn] = React.useState<UnregisterCallback | null>(null);
...
    if (unblockFn) {
      unblockFn();
      setUnblockFn(null);
    }
...
const unBlockHandler = history.block((prompt) => {...
setUnblockFn(unBlockHandler);
```

But unblockFn is always null in that case.

Reply:
I was also trying that and getting frustrated. Something dumb about closures and function scope. I think it's because the useEffect cleanup function has to reference `unblockFn` from the function scope where the cleanup function was declared, which is the scope from the render before `setUnblockFn` was called. Since you're using `unblockFn` inside the useEffect it is in the dependencies for the effect, which means when you `setUnblockFn` it will cause a re-render with a new `unblockFn`, so the effect will cleanup before that render, before it has access to the new `unblockFn`. 

Anyway, since we don't need to render any JSX differently based on the unblockFn, putting it in a useState is sorta overkill. It's the same reason you put return values of setTimeout / setInterval in refs, they need to sort of exist as a persistent thing outside the render lifecycle.

My brain hurts.

Reply:
Thx for explaining! I was speculating something of the sort, just wasn't so confident so wanted to get your take on it. Nicely done, sir.